pkg/fswatcher/fswatcher.go

```diff
@@ -64,285 +89,209 @@ type Watcher struct {
-	// If the path is already watched, we do not want to add it to fsnotify
-	// again, thus the check on the refcount first.
-	// Note: If we already watchedPath has been invalidated recently,
-	// this if statement will be false (because invalidateWatch resets the
-	// count)
-	if w.watchedPathCount[path] > 0 {
-		w.watchedPathCount.Add(path)
-		return path, nil
-	}
+func (w *Watcher) loop() {
+	defer w.wg.Done()
 
-	// Adds the file to fsnotify. Important note: If path is a symlink, this
-	// will watch the *target* of the symlink. So any event we will observe,
-	// will be valid for the target, not for the symlink itself. The reported
-	// path in the events however will remain the path of the symlink.
-	err := w.watcher.Add(path)
-	if err != nil {
-		// if the path does not exist, try to watch its parent instead
-		if errors.Is(err, os.ErrNotExist) {
-			parent := filepath.Dir(path)
-			if parent != path {
-				return w.startWatching(parent)
-			}
+	for {
+		select {
+		case <-w.ticker:
+			w.tick()
+		case <-w.stop:
+			return
 		}
-
-		return "", err
-	}
-
-	// Start counting the references for the watched path.
-	// The following is identical to `w.watchedPathCount[path] = 1`, because
-	// w.watchedPathCount[path] was zero when we entered the function
-	w.watchedPathCount.Add(path)
-	return path, nil
-}
-
-func (w *Watcher) stopWatching(path string) {
-	// Decrease the refcount for the old watchedPath. If this was the last
-	// use of this watchedPath, we remove it from the underlying fsnotify
-	// watcher.
-	if w.watchedPathCount.Delete(path) {
-		_ = w.watcher.Remove(path)
 	}
 }
 
-func (w *Watcher) invalidateWatch(path string) {
-	if w.watchedPathCount[path] > 0 {
-		delete(w.watchedPathCount, path)
-		// The result is ignored because fsnotify removes deleted paths by
-		// itself, in which case it will complain about a non-existing path
-		// being removed.
-		_ = w.watcher.Remove(path)
-	}
-}
-
-// hasParent returns true if path is a child or equal to parent
-func hasParent(path, parent string) bool {
-	path = filepath.Clean(path)
-	parent = filepath.Clean(parent)
-	if path == parent {
-		return true
-	}
+func (w *Watcher) tick() {
+	for _, oldState := range w.tracked {
+		path := oldState.path
+		oldInfo := oldState.info
+		newState := state{path: oldState.path}
+
+		// os.Stat follows symlinks, os.Lstat doesn't
+		info, err := os.Lstat(path)
+		newState.info = info
+
+		if os.IsNotExist(err) {
+			// if the path does not exist, check if it existed before because if it
+			// did -- issue a deletion event
+			if oldState.info != nil {
+				// this file was deleted
+				w.sendEvent(Event{
+					Name: path,
+					Op:   Remove,
+				})
+
+				// clear out old state from the map
+				w.tracked[oldState.path] = newState
+			}
 
-	for {
-		pathParent := filepath.Dir(path)
-		if pathParent == parent {
-			return true
+			// the path doesn't exist, and it didn't exist before -- ignore for now
```

Comment:
We're still going through this path when the file did exist before and was removed. As-is it's easy enough to understand because the block is small and granted it's not expected to grow so ok.

Reply:
Yeah I think this comment is just incorrect as a result of a refactor at some point